src/shared/components/home/tagline-form.tsx

```diff
@@ -99,65 +158,153 @@ export class TaglineForm extends Component<TaglineFormProps, TaglineFormState> {
-                {this.props.loading ? (
+                {this.state.loading ? (
                   <Spinner />
                 ) : (
                   capitalizeFirstLetter(I18NextService.i18n.t("save"))
                 )}
               </button>
+              {this.hasPendingChanges() && (
+                <button
+                  onClick={linkEvent(this, this.handleCancelClick)}
+                  className="btn btn-secondary me-2"
+                >
+                  {I18NextService.i18n.t("cancel")}
+                </button>
+              )}
             </div>
           </div>
+          <div>
+            <Paginator
+              page={this.state.page}
+              onChange={this.handlePageChange}
+              nextDisabled={false}
+              disabled={this.hasPendingChanges()}
+            />
+          </div>
         </div>
       </div>
     );
   }
 
   handleTaglineChange(i: TaglineForm, index: number, val: string) {
-    if (i.state.taglines) {
-      i.setState(prev => ({
-        ...prev,
-        taglines: prev.taglines.map((tl, i) => (i === index ? val : tl)),
-      }));
-    }
+    const editable = i.state.taglines[index];
+    i.setState(() => {
+      markForUpdate(editable);
+      const tagline: Tagline = editable.tagline;
+      tagline.content = val;
+    });
   }
 
-  handleDeleteTaglineClick(d: { i: TaglineForm; index: number }, event: any) {
+  async handleDeleteTaglineClick(
+    d: { i: TaglineForm; index: number },
+    event: any,
+  ) {
     event.preventDefault();
-    d.i.setState(prev => ({
-      ...prev,
-      taglines: prev.taglines.filter((_, i) => i !== d.index),
-      editingRow: undefined,
-    }));
+    const editable = d.i.state.taglines[d.index];
+    if (editable.change === "create") {
+      // This drops the entry immediately, other deletes have to be saved.
+      d.i.setState(prev => {
+        return { taglines: prev.taglines.filter(x => x !== editable) };
+      });
+    } else {
+      d.i.setState(() => {
+        editable.change = "delete";
+        editable.editMode = false;
+      });
+    }
   }
 
   handleEditTaglineClick(d: { i: TaglineForm; index: number }, event: any) {
     event.preventDefault();
-    if (d.i.state.editingRow === d.index) {
-      d.i.setState({ editingRow: undefined });
-    } else {
-      d.i.setState({ editingRow: d.index });
-    }
+    const editable = d.i.state.taglines[d.index];
+    d.i.setState(prev => {
+      prev.taglines
+        .filter(x => x !== editable)
+        .forEach(x => {
+          x.editMode = false;
+        });
+      editable.editMode = !editable.editMode;
+    });
   }
 
   async handleSaveClick(i: TaglineForm) {
-    i.props.onSaveSite({
-      taglines: i.state.taglines,
-    });
+    const promises: Promise<any>[] = [];
+    for (const editable of i.state.taglines) {
+      if (editable.change === "update") {
+        promises.push(
+          HttpService.client.editTagline(editable.tagline).then(res => {
```

Comment:
This one to me doesn't seem like the right way to do it. here's

Admin settings screen -> Taglines component -> click to edit Tagline form.

The tagline form then pushes up (IE `onEdit(form: TaglineForm)`, or `onCreate(form: ..)` up the chain, up to the admin settings screen, which is the only one that should be making HTTP calls.

Reply:
The only downside to the way it is now, is that SSR doesn't render any actual data. But the emoji and tagline tabs are not visible until opening the tab. Moving the HTTP calls would only create properties with same signatures as the API methods. This way also the state for AdminSettings doesn't have to change, which means only one child has to rerender instead of all of them.

Reply:
That's true. Its not a blocker for this anyway, and we can always try to move these http calls to the top level screens later.